File: foolish-core-java/src/main/java/org/foolish/fvm/ubc/IdentifierFiroe.java
package org.foolish.fvm.ubc;

import org.foolish.ast.AST;
/**
 * IdentifierFiroe represents a characterized identifier reference in the UBC system.
 *
 * Uses CharacterizedIdentifier to hold the identifier name and characterization:
 * - Simple identifier: `x` → id="x", characterization=""
 * - Characterized: `type'x` → id="x", characterization="type"
 *
 * The characterization is used when resolving identifiers to:
 * - Disambiguate between multiple bindings of the same name
 * - Type checking (future)
 * - Scope resolution (future)
 *
 * Currently, identifier lookup is not yet implemented in UBC, so this
 * returns NK (not-known) values.
 */

public class IdentifierFiroe extends FiroeWithBraneMind implements Constanicable {
    private final Query.StrictlyMatchingQuery identifier;
    FIR value = null; // Package-private for access by RegexpSearchFiroe

    public IdentifierFiroe(AST.Identifier identifier) {
        super(identifier);
        this.identifier = new Query.StrictlyMatchingQuery(identifier.id(), identifier.canonicalCharacterization());
    }

    /**
     * Copy constructor for cloneConstanic.
     */
    protected IdentifierFiroe(IdentifierFiroe original, FIR newParent) {
        super(original, newParent);
        this.identifier = original.identifier;  // Query is immutable, can share
        this.value = null;  // Reset value for re-resolution in new context
    }

    /**
     * Gets the CharacterizedIdentifier.
     */
    public CharacterizedIdentifier getIdentifier() {
        return identifier;
    }

    /**
     * Gets the characterization as a flattened string.
     * Returns empty string "" if no characterization.
     */
    public String getCharacterization() {
        return identifier.getCharacterization();
    }

    /**
     * Gets the resolved FIR value for this identifier.
     * Returns null if not yet resolved or if resolution failed.
     */
    public FIR getResolvedFir() {
        return value;
    }

    @Override
    public FIR getResult() {
        return value;
    }

    /**
     * An identifier is Constanic if it hasn't been resolved yet or if its resolved value is Constanic.
     */
    @Override
    public boolean isConstanic() {
        if (value == null) {
            return true; // Not yet resolved or missing
        }
        return value.isConstanic();
    }

    @Override
    protected void initialize() {
        setInitialized();
    }

    /**
     * Implement the step method only overriding resolution phase, during resolution
     * we use the branemind to find the value of the identifier and store a reference to
     * it for `getValue()`
     */
    @Override
    public int step() {
        System.err.println("DEBUG STEP: Identifier " + identifier.getId() + " step() at state " + getNyes());
        switch (getNyes()) {
            case UNINITIALIZED -> {
                initialize();
                setNyes(Nyes.INITIALIZED);
                return 1;
            }
            case INITIALIZED -> {
                var found = memoryGet(identifier, 0);
                if (found.isEmpty()) {
                    System.err.println("DEBUG: Identifier " + identifier.getId() + " NOT FOUND");
                    setNyes(Nyes.CONSTANIC);
                    return 1;
                }
                value = found
                        .map(r -> r.getRight())
                        .orElse(null);
                if (value == null) {
                    setNyes(Nyes.CONSTANIC);
                } else {
                    System.err.println("DEBUG: Identifier " + identifier.getId() + " FOUND, transitioning to CHECKED");
                    setNyes(Nyes.CHECKED);
                }
                return 1;
            }
            case CHECKED -> {
                // Identifier lookup is complete, check if value has reached final state
                // Use atConstanic() to check for exactly CONSTANIC (unresolved)
                // Use atConstant() to check for exactly CONSTANT (fully resolved)
                System.err.println("DEBUG: Identifier " + identifier.getId() + " CHECKED value=" +
                    value.getClass().getSimpleName() + " atConstanic=" + value.atConstanic() +
                    " atConstant=" + value.atConstant() + " nyes=" + value.getNyes() + " isNye=" + value.isNye());
                if (value.atConstanic()) {
                    // CONSTANIC value needs to be coordinated (cloned and re-positioned)
                    // so it can re-resolve from this identifier's position.
                    System.err.println("DEBUG: Identifier " + identifier.getId() + " coordinating CONSTANIC value " + value.getClass().getSimpleName());
                    // Clone the value with INITIALIZED state so it will re-evaluate.
                    value = value.cloneConstanic(this, java.util.Optional.of(Nyes.INITIALIZED));
                    // storeFirs will ordinate the clone to this identifier's context
                    storeFirs(value);
                    setNyes(Nyes.PRIMED);
                } else if (value.atConstant()) {
                    setNyes(Nyes.CONSTANT);
                } else {
                    // Value is still evaluating - shouldn't happen for identifiers
                    // since we just store a reference, but handle gracefully
                    setNyes(Nyes.EVALUATING);
                }
                return 1;
            }
            case PRIMED -> {
                // Transition to EVALUATING to step the coordinated value
                setNyes(Nyes.EVALUATING);
                return 1;
            }
            case EVALUATING -> {
                // Step the coordinated value through evaluation
                if (isBraneEmpty()) {
                    // Value finished evaluating, check final state
                    if (value.atConstanic()) {
                        setNyes(Nyes.CONSTANIC);
                    } else {
                        setNyes(Nyes.CONSTANT);
                    }
                    return 1;
                }
                // Step the next FIR in braneMind
                FIR current = braneDequeue();
                current.step();
                if (current.isNye()) {
                    braneEnqueue(current);
                }
                return 1;
            }
            case CONSTANIC, CONSTANT -> {
                return 0;
            }
            default -> {
                return super.step();
            }
        }
    }
    /**
     * Identifier lookup is not yet implemented.
     * @throws UnsupportedOperationException always
     */
    @Override
    public long getValue() {
        if (value == null) {
            if (atConstanic()) throw new IllegalStateException("Identifier is Constanic (missing): " + identifier.getId());
            throw new IllegalStateException("Identifier not resolved or not found: " + identifier.getId());
        }
        return value.getValue();
    }

    @Override
    public String toString() {
        return ((AST.Identifier) ast).toString();
    }

    @Override
    protected FIR cloneConstanic(FIR newParent, java.util.Optional<Nyes> targetNyes) {
        if (!isConstanic()) {
            throw new IllegalStateException(
                formatErrorMessage("cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
                                  "but this FIR is in state: " + getNyes()));
        }

        if (isConstant()) {
            return this;  // Share CONSTANT identifiers completely
        }

        // CONSTANIC: use copy constructor
        IdentifierFiroe copy = new IdentifierFiroe(this, newParent);

        // Set target state if specified, otherwise copy from original
        if (targetNyes.isPresent()) {
            copy.nyes = targetNyes.get();
        } else {
            copy.nyes = this.nyes;
        }

        return copy;
    }
}

File: foolish-core-scala/src/main/scala/org/foolish/fvm/scubc/IdentifierFiroe.scala
package org.foolish.fvm.scubc

import org.foolish.ast.AST
import org.foolish.fvm.scubc.BraneMemory.StrictlyMatchingQuery

/**
 * IdentifierFiroe represents a characterized identifier reference in the UBC system.
 *
 * Uses CharacterizedIdentifier to hold the identifier name and characterization:
 * - Simple identifier: `x` → id="x", characterization=""
 * - Characterized: `type'x` → id="x", characterization="type"
 *
 * The characterization is used when resolving identifiers to:
 * - Disambiguate between multiple bindings of the same name
 * - Type checking (future)
 * - Scope resolution (future)
 */
class IdentifierFiroe(override val ast: AST.Identifier)
  extends FiroeWithBraneMind(ast) with Constanicable:

  private val identifier = new StrictlyMatchingQuery(
    ast.id(),
    ast.canonicalCharacterization()
  )
  private[scubc] var value: FIR = null // Package-private for access by RegexpSearchFiroe

  def this(id: String, characterization: String) =
    this(new AST.Identifier(
      if characterization == null || characterization.isEmpty then java.util.List.of() else java.util.List.of(characterization),
      id
    ))

  /** Gets the CharacterizedIdentifier */
  def getIdentifier: CharacterizedIdentifier = identifier

  /** Gets the characterization as a flattened string */
  def getCharacterization: String = identifier.getCharacterization

  /**
   * An identifier is abstract if it hasn't been resolved yet or if its resolved value is abstract.
   */
  override def isAbstract: Boolean =
    if atConstanic then
      true
    else if value == null then
      true // Not yet resolved
    else
      value.isAbstract

  /**
   * An identifier is Constanic only when its state is CONSTANIC or CONSTANT.
   * A null value in earlier states means the identifier hasn't been evaluated yet.
   */
  override def isConstanic: Boolean =
    if getNyes != Nyes.CONSTANIC && getNyes != Nyes.CONSTANT then
      false
    else if value == null then
      true  // Not yet resolved or out of bounds
    else
      value.isConstanic

  override protected def initialize(): Unit = ()

  /**
   * Implement the step method, during checking phase we use the branemind
   * to find the value of the identifier and store a reference to it for getValue()
   */
  override def step(): Int =
    getNyes match
      case Nyes.INITIALIZED =>
        val parentMem = braneMemory.getParent
        val parentStr = if parentMem != null then
          s"parentSize=${parentMem.size} parentHash=${System.identityHashCode(parentMem)}"
        else "noParent"
        System.out.println(s"DEBUG IdentifierFiroe.step INITIALIZED: this=${System.identityHashCode(this)} identifier=$identifier braneMemory=${braneMemory.size} $parentStr")
        value = braneMemory.get(identifier, 0)
          .map(_._2)
          .orNull
        System.out.println(s"DEBUG IdentifierFiroe.step INITIALIZED: value=$value valueClass=${if (value != null) value.getClass.getSimpleName else "null"}")
        if value == null then
          setNyes(Nyes.CONSTANIC)
        else
          setNyes(Nyes.CHECKED)
        1
      case Nyes.CHECKED =>
        if value.isConstanic then
          // For CONSTANIC values, we DON'T clone them here.
          // Instead, we keep a reference to the original value and let the
          // parent (e.g., ConcatenationFiroe in performJoin) handle the cloning.
          // This ensures that when the value is cloned, its parent chain is set
          // correctly (to the ConcatenationFiroe, not the IdentifierFiroe).
          //
          // We still store the value in braneMemory for unwrapConstanicable to work,
          // but we don't enqueue it for re-evaluation (it will be re-evaluated after
          // the parent clones it).
          storeFirs(value)
          setNyes(Nyes.PRIMED)
        else if value.isConstant then
          storeFirs(value)
          setNyes(Nyes.CONSTANT)
        else
          storeFirs(value)
          braneEnqueue(value)
          setNyes(Nyes.PRIMED)
        1
      case _ =>
        super.step()

  /**
   * Get the value of the resolved identifier.
   */
  override def getValue: Long =
    if atConstanic then
      throw UnsupportedOperationException("Cannot get value from constanic identifier")
    value.getValue

  /** Returns the resolved value for unwrapping in search operations */
  override def getResult: FIR = value


  override def valuableSelf(): java.util.Optional[FIR] =
    if value == null then
      if atConstanic then
        java.util.Optional.empty[FIR]()
      else
        null  // Not ready yet
    else if atConstanic then
      java.util.Optional.empty[FIR]()
    else
      value.valuableSelf()

  override def cloneConstanic(newParent: FIR, targetNyes: Option[Nyes]): FIR =
    if !isConstanic then
      throw IllegalStateException(
        s"cloneConstanic can only be called on CONSTANIC or CONSTANT FIRs, " +
        s"but this FIR is in state: ${getNyes}")
    if isConstant then
      return this  // Share CONSTANT identifiers
    // CONSTANIC: create fresh copy from AST
    val copy = new IdentifierFiroe(ast.asInstanceOf[AST.Identifier])
    copy.setParentFir(newParent)
    copy.value = null  // Reset for re-evaluation
    copy.setNyes(getNyes)
    // Set target state if specified
    targetNyes.foreach(copy.setNyes)
    // Ordinate to parent so identifier resolution works correctly
    // This is critical for cloned identifiers to find their values in the new context
    if !copy.ordinated then
      copy.ordinateToParentBraneMind(newParent.asInstanceOf[FiroeWithBraneMind])
    copy

  override def toString: String = ast.toString
